File: explore/explore_routes.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.9"
# dependencies = []
# ///
"""List SF Bay Ferry routes, and dump the stops for a route.

    uv run explore_routes.py                  # list all routes
    uv run explore_routes.py --route 19417    # list stops on route 19417

Use this to discover the stop_id values for Seaplane Lagoon and the
SF Ferry Building, then hard-code them into trip_updates.py.
"""

import argparse

from swiftly import AGENCY_KEY, get_json, pick


def find_stops(node, found):
    """Recursively collect stop-like dicts (have an id and a lat/lon)."""
    if isinstance(node, dict):
        sid = pick(node, "id")
        lat = pick(node, "lat", "latitude")
        lon = pick(node, "lon", "longitude")
        if sid is not None and lat is not None and lon is not None:
            found[str(sid)] = {
                "id": str(sid),
                "name": pick(node, "name", "title") or "(unnamed)",
                "lat": lat,
                "lon": lon,
            }
        for value in node.values():
            find_stops(value, found)
    elif isinstance(node, list):
        for item in node:
            find_stops(item, found)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--route", help="route id; dumps that route's stops")
    args = parser.parse_args()

    if args.route:
        data = get_json(
            f"/info/{AGENCY_KEY}/routes", route=args.route, verbose="true", format="json"
        )
        routes = pick(data.get("data", {}), "routes") or []
        if not routes:
            print(f"No route {args.route} found.")
            return
        for route in routes:
            print(f"\nRoute {pick(route, 'id')}  {pick(route, 'name', 'longName')}")
            stops = {}
            find_stops(route, stops)
            for s in sorted(stops.values(), key=lambda x: x["name"]):
                print(f"  stop_id={s['id']:<10} {s['name']:<32} ({s['lat']}, {s['lon']})")
    else:
        data = get_json(f"/info/{AGENCY_KEY}/routes", format="json")
        routes = pick(data.get("data", {}), "routes") or []
        print(f"{len(routes)} routes for agency '{AGENCY_KEY}':\n")
        for route in sorted(routes, key=lambda r: str(pick(r, "id"))):
            print(
                f"  id={str(pick(route, 'id')):<10} "
                f"{pick(route, 'shortName') or '':<8} "
                f"{pick(route, 'longName') or pick(route, 'name') or ''}"
            )
        print("\nNext: uv run explore_routes.py --route <id>  to see its stops.")


if __name__ == "__main__":
    main()

File: scripts/explore-routes.mjs
#!/usr/bin/env node
// List SF Bay Ferry routes, and dump the stops for a route.
//
//   node scripts/explore-routes.mjs                  # list all routes
//   node scripts/explore-routes.mjs --route 19417    # list stops on route 19417
//
// Use this to discover stop ids, or — if you're installing the TRMNL
// plugin — to find your agency's numeric route id for the form's Route
// field.

import { parseArgs } from 'node:util';
import { AGENCY_KEY, getJson, pick } from './swiftly.mjs';

const { values } = parseArgs({ options: { route: { type: 'string' } } });

// Recursively collect stop-like objects (have an id and a lat/lon).
function findStops(node, found) {
  if (Array.isArray(node)) {
    for (const item of node) findStops(item, found);
    return;
  }
  if (!node || typeof node !== 'object') return;

  const id = pick(node, 'id');
  const lat = pick(node, 'lat', 'latitude');
  const lon = pick(node, 'lon', 'longitude');
  if (id != null && lat != null && lon != null) {
    found.set(String(id), {
      id: String(id),
      name: pick(node, 'name', 'title') || '(unnamed)',
      lat,
      lon,
    });
  }
  for (const value of Object.values(node)) findStops(value, found);
}

if (values.route) {
  const data = await getJson(`/info/${AGENCY_KEY}/routes`, {
    route: values.route,
    verbose: 'true',
    format: 'json',
  });
  const routes = pick(data.data ?? {}, 'routes') ?? [];
  if (routes.length === 0) {
    console.log(`No route ${values.route} found.`);
    process.exit(0);
  }
  for (const route of routes) {
    console.log(`\nRoute ${pick(route, 'id')}  ${pick(route, 'name', 'longName')}`);
    const stops = new Map();
    findStops(route, stops);
    for (const s of [...stops.values()].sort((a, b) => a.name.localeCompare(b.name))) {
      console.log(`  stop_id=${s.id.padEnd(10)} ${s.name.padEnd(32)} (${s.lat}, ${s.lon})`);
    }
  }
} else {
  const data = await getJson(`/info/${AGENCY_KEY}/routes`, { format: 'json' });
  const routes = pick(data.data ?? {}, 'routes') ?? [];
  console.log(`${routes.length} routes for agency '${AGENCY_KEY}':\n`);
  for (const route of [...routes].sort((a, b) =>
    String(pick(a, 'id')).localeCompare(String(pick(b, 'id'))),
  )) {
    console.log(
      `  id=${String(pick(route, 'id')).padEnd(10)} ` +
        `${(pick(route, 'shortName') || '').padEnd(8)} ` +
        `${pick(route, 'longName') || pick(route, 'name') || ''}`,
    );
  }
  console.log('\nNext: node scripts/explore-routes.mjs --route <id>  to see its stops.');
}
